firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java

```diff
@@ -191,51 +213,58 @@ public synchronized void onActivityResumed(Activity activity) {
     appStartActivity = new WeakReference<Activity>(activity);
 
     onResumeTime = clock.getTime();
-    final Timer startTime = FirebasePerfProvider.getAppStartTime();
+    this.appStartTime = FirebasePerfProvider.getAppStartTime();
     AndroidLogger.getInstance()
         .debug(
             "onResume(): "
                 + activity.getClass().getName()
                 + ": "
-                + startTime.getDurationMicros(onResumeTime)
+                + this.appStartTime.getDurationMicros(onResumeTime)
                 + " microseconds");
 
+    // Log the app start trace in a non-main thread.
+    executorService.execute(this::logAppStartTrace);
+
+    if (isRegisteredForLifecycleCallbacks) {
+      // After AppStart trace is logged, we can unregister this callback.
+      unregisterActivityLifecycleCallbacks();
+    }
+  }
+
+  private void logAppStartTrace() {
     TraceMetric.Builder metric =
         TraceMetric.newBuilder()
             .setName(Constants.TraceNames.APP_START_TRACE_NAME.toString())
-            .setClientStartTimeUs(startTime.getMicros())
-            .setDurationUs(startTime.getDurationMicros(onResumeTime));
+            .setClientStartTimeUs(getappStartTime().getMicros())
+            .setDurationUs(getappStartTime().getDurationMicros(onResumeTime));
     List<TraceMetric> subtraces = new ArrayList<>(/* initialCapacity= */ 3);
 
-    TraceMetric.Builder temp =
+    TraceMetric.Builder traceMetricBuilder =
         TraceMetric.newBuilder()
             .setName(Constants.TraceNames.ON_CREATE_TRACE_NAME.toString())
-            .setClientStartTimeUs(startTime.getMicros())
-            .setDurationUs(startTime.getDurationMicros(onCreateTime));
-    subtraces.add(temp.build());
+            .setClientStartTimeUs(getappStartTime().getMicros())
+            .setDurationUs(getappStartTime().getDurationMicros(onCreateTime));
+    subtraces.add(traceMetricBuilder.build());
 
-    temp = TraceMetric.newBuilder();
-    temp.setName(Constants.TraceNames.ON_START_TRACE_NAME.toString())
+    traceMetricBuilder = TraceMetric.newBuilder();
+    traceMetricBuilder
+        .setName(Constants.TraceNames.ON_START_TRACE_NAME.toString())
         .setClientStartTimeUs(onCreateTime.getMicros())
         .setDurationUs(onCreateTime.getDurationMicros(onStartTime));
-    subtraces.add(temp.build());
+    subtraces.add(traceMetricBuilder.build());
 
-    temp = TraceMetric.newBuilder();
-    temp.setName(Constants.TraceNames.ON_RESUME_TRACE_NAME.toString())
+    traceMetricBuilder = TraceMetric.newBuilder();
+    traceMetricBuilder
+        .setName(Constants.TraceNames.ON_RESUME_TRACE_NAME.toString())
         .setClientStartTimeUs(onStartTime.getMicros())
         .setDurationUs(onStartTime.getDurationMicros(onResumeTime));
-    subtraces.add(temp.build());
+    subtraces.add(traceMetricBuilder.build());
 
     metric
         .addAllSubtraces(subtraces)
         .addPerfSessions(SessionManager.getInstance().perfSession().build());
```

Comment:
is this thread-safe? Would it make sense to add `synchronized` to `SessionManager.getInstance()`? But `transportManager.log` 2 lines down also calls `SessionManager.getInstance()` in a different thread already, so I'm not sure if it's already thread-safe.

Reply:
Good catch! I stored SessionManager.getInstance().perfSession() into a class variable.